### PR TITLE
[FLASK] Add update snap UI

### PR DESCRIPTION
**Pull request by hmalik88**
Enables dapp-triggered updates for snaps. A permission list is displayed showing new permissions, approved permissions and revoked permissions. Going through this connect flow will ensure that a snap is updated. Future iterations to the update snap flow will include user-triggered updates and rollbacks.

[Figma design](https://www.figma.com/file/mKIkB5V0k2iYUpz7w7syGO/Update-Snap?node-id=444%3A1077) 

Steps to reproduce:

1. Switch to this branch and run a flask build using `yarn start --build-type flask` 
2. Make sure this version of the extension is added to chrome.
3. You can write a simple server to serve an index.html file that does `ethereum.request` OR use the snaps-cli and do `mm-snap init` to scaffold a template project from which you can edit the `connect` function (and subsequently run `mm-snap build && mm-snap serve`.
4. I've published a test snap with two versions that you can test against: update-test-snap [v1.0.0, v1.0.1].

Screen-recording: 

https://user-images.githubusercontent.com/41640681/177984039-e7ca2eae-ca60-4578-b70d-0da7d94433c8.mov

**Note**: E2E test is forthcoming.

### CODE REVIEW

File: app/_locales/en/messages.json
Comment (FrederikBolding) [inline app/_locales/en/messages.json:293]: I think you need to describe what `$1` is supposed to be replaced by.

File: app/_locales/en/messages.json
Comment (FrederikBolding) [inline app/_locales/en/messages.json:3018]: I think you need to describe what `$1` is supposed to be replaced by.

File: ui/components/app/flask/snaps-authorship-pill/snaps-authorship-pill.js
Comment (FrederikBolding) [inline ui/components/app/flask/snaps-authorship-pill/snaps-authorship-pill.js:16]: Why is this needed?

File: ui/components/app/flask/update-snap-permission-list/update-snap-permission-list.js
Comment (FrederikBolding) [inline ui/components/app/flask/update-snap-permission-list/update-snap-permission-list.js:27]: I think we should use `luxon` for date formatting. We do this elsewhere as well.

File: ui/components/app/permissions-connect-header/permissions-connect-header.component.js
Comment (FrederikBolding) [inline ui/components/app/permissions-connect-header/permissions-connect-header.component.js:30]: Do we need two props for this? Could we just rename the existing one?

File: ui/pages/home/home.container.js
Comment (FrederikBolding) [inline ui/pages/home/home.container.js:95]: What is the goal of this? 🤔

File: ui/pages/permissions-connect/permissions-connect.container.js
Comment (FrederikBolding) [inline ui/pages/permissions-connect/permissions-connect.container.js:60]: It would be preferable if we could do this in a cleaner way, I'm not entirely sure how though 🤔

File: ui/pages/permissions-connect/permissions-connect.component.js
Comment (FrederikBolding) [inline ui/pages/permissions-connect/permissions-connect.component.js:121]: Couldn't this be detected by the fact that the snap is already installed?

Comment (FrederikBolding) [reply ui/pages/permissions-connect/permissions-connect.component.js:121]: Otherwise, maybe we need an enum instead of a flag for future use

File: ui/pages/permissions-connect/flask/snap-update/snap-update.js
Comment (FrederikBolding) [inline ui/pages/permissions-connect/flask/snap-update/snap-update.js:111]: Is this equal to the NPM link or where does it come from?

File: ui/pages/permissions-connect/permissions-connect.container.js
Comment (FrederikBolding) [reply ui/pages/permissions-connect/permissions-connect.container.js:60]: Is this what you were saying about the shapes of the objects not matching up entirely?

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

File: ui/components/app/flask/snaps-authorship-pill/snaps-authorship-pill.js
Comment (hmalik88) [reply ui/components/app/flask/snaps-authorship-pill/snaps-authorship-pill.js:16]: Leads to a type error when trying to access undefined for `.startsWith` and `.replace`, but I suppose I can just use optional chaining.

Comment (FrederikBolding) [reply ui/components/app/flask/snaps-authorship-pill/snaps-authorship-pill.js:16]: But it should be defined, no?

File: ui/pages/home/home.container.js
Comment (hmalik88) [reply ui/pages/home/home.container.js:95]: The home component grabs whatever pending approvals there are and sends them to the `PermissionConnect` component, in the current case, `firstPermissionRequest` would not pick up the pending approval because it's not checking for update requests.

**Comment by FrederikBolding** [comment]
E2Es are failing because some code is not fenced correctly:

```
Uncaught TypeError: (0 , _selectors.getFirstSnapUpdateRequest) is not a function
  at Function.mapStateToProps [as mapToProps] (home.container.js:93:31)
```

File: ui/components/app/permissions-connect-header/permissions-connect-header.component.js
Comment (hmalik88) [reply ui/components/app/permissions-connect-header/permissions-connect-header.component.js:30]: Yeah, true.

File: ui/pages/permissions-connect/permissions-connect.container.js
Comment (hmalik88) [reply ui/pages/permissions-connect/permissions-connect.container.js:60]: Yessir! We can make it clean again by having the new methods I suggested 😄

File: ui/pages/permissions-connect/permissions-connect.component.js
Comment (hmalik88) [reply ui/pages/permissions-connect/permissions-connect.component.js:121]: hmm yeah or I can just check for `permissionsRequest.newPermissions`

File: ui/components/app/flask/snaps-authorship-pill/snaps-authorship-pill.js
Comment (hmalik88) [reply ui/components/app/flask/snaps-authorship-pill/snaps-authorship-pill.js]: It should, but it throws if you don't. Specifically, when you hit cancel or approve.

File: ui/pages/permissions-connect/flask/snap-update/index.scss
Comment (hmalik88) [reply ui/pages/permissions-connect/flask/snap-update/index.scss:32]: But then it's not consistent across install and update screens, I fixed the wrapping issue with this padding rule.

File: ui/pages/permissions-connect/permissions-connect.container.js
Comment (FrederikBolding) [reply ui/pages/permissions-connect/permissions-connect.container.js:59]: I believe we can fix this by doing something like this: https://github.com/MetaMask/snaps-skunkworks/pull/614

**Review by FrederikBolding — DISMISSED** [review]

File: ui/pages/permissions-connect/flask/snap-update/snap-update.js
Comment (hmalik88) [reply ui/pages/permissions-connect/flask/snap-update/snap-update.js:81]: I saw Mark Stacey using this pattern, I don't think I had a good reason why.

File: ui/components/app/flask/snaps-authorship-pill/snaps-authorship-pill.js
Comment (hmalik88) [reply ui/components/app/flask/snaps-authorship-pill/snaps-authorship-pill.js:19]: Ok, I think I know what the issue is. There's a race condition if I'm not mistaken, when you hit cancel the request gets deleted and hence there is no reference to request.snapId. Things work with snap-install because when installing a snap for the first time, the metadata is added before the permissions are requested so we have access to snapId directly from persisted state, whereas when you hit cancel the request that supplies the snapId is being deleted right away. I mentioned this issue in my slack post, we can leave this like it is and I can take care of it in a subsequent PR.

File: ui/pages/permissions-connect/permissions-connect.container.js
Comment (hmalik88) [reply ui/pages/permissions-connect/permissions-connect.container.js:59]: Sorry, in this case the shape of the object won't matter because we still need to call `getSnapUpdateRequests` since the `getPermissionsRequests` selector does not check for updates.

Comment (hmalik88) [reply ui/pages/permissions-connect/permissions-connect.container.js:59]: Actually, I'll write a selector that just grabs both.

File: ui/pages/permissions-connect/flask/snap-update/snap-update.js
Comment (hmalik88) [reply ui/pages/permissions-connect/flask/snap-update/snap-update.js:103]: removing from the `SnapInstall` component as well then.

**Review by georgewrmarshall — COMMENTED** [review]
UI looks good to me. Checked the font wrapping issue on the button and it looks like it's been resolved with the adjusted button padding. A couple nits but nothing blocking merge

File: ui/pages/permissions-connect/permissions-connect.component.js
Comment (hmalik88) [reply ui/pages/permissions-connect/permissions-connect.component.js:119]: It would, but I'm not so keen on updating existing extension selectors to have snaps logic. I think it would make sense to do that once snaps hits main extension release. I've left a comment in the code to update when that happens.

**Review by ritave — DISMISSED** [review]

**Review by brad-decker — APPROVED** [review]
LGTM

**Review by ritave — APPROVED** [review]

**Comment by hmalik88** [comment]
Update snaps functionality is blocked by this bug that was just uncovered: [Fix Update Snap Request Object](https://github.com/MetaMask/snaps-skunkworks/issues/672)

**Pull request merged**
